In [2]:
import pandas as pd

In [3]:
customer_df = pd.read_csv("../dataset/Customer.csv")
prod_cat_df = pd.read_csv("../dataset/prod_cat_info.csv")
trans_df = pd.read_csv("../dataset/Transactions.csv")

In [5]:
print("Customer_df shape:", customer_df.shape)
print("Product_Category_df shape:", prod_cat_df.shape)
print("Transactions_df shape:", trans_df.shape)

print("Number of Unique Customers:", customer_df['customer_Id'].nunique())
print("Number of Unique Products:", prod_cat_df['prod_cat_code'].nunique())
print("Number of Unique Categories:", prod_cat_df['prod_cat'].nunique())
print("Number of Unique Sub-Categories:", prod_cat_df['prod_sub_cat_code'].nunique())
print("Number of Unique Transactions:", trans_df['transaction_id'].nunique())

Customer_df shape: (5647, 4)
Product_Category_df shape: (23, 4)
Transactions_df shape: (23053, 10)
Number of Unique Customers: 5647
Number of Unique Products: 6
Number of Unique Categories: 6
Number of Unique Sub-Categories: 12
Number of Unique Transactions: 20878


In [6]:
customer_df.isnull().sum()

customer_Id    0
DOB            0
Gender         2
city_code      2
dtype: int64

In [19]:
# showing the null values in the customer_df dataset. When 'Gender' or 'city_code' is null.
customer_null = customer_df[customer_df['Gender'].isnull() | customer_df['city_code'].isnull()]
customer_null.head()

,customer_Id,DOB,Gender,city_code
24,267199,14-02-1970,NaN,2.0
87,271626,02-06-1970,NaN,6.0
115,268447,14-07-1970,M,NaN
149,268709,09-09-1970,F,NaN


In [33]:
print("Min / Youngest Customer Age:", customer_df['DOB'].max())
print("Max / Oldest Customer Age:", customer_df['DOB'].min())

Min / Youngest Customer Age: 29-12-1992
Max / Oldest Customer Age: 01-01-1972


In [34]:
print("Number of unique city_codes:", customer_df['city_code'].nunique())

Number of unique city_codes: 10


In [7]:
prod_cat_df.isnull().sum()

prod_cat_code        0
prod_cat             0
prod_sub_cat_code    0
prod_subcat          0
dtype: int64

In [20]:
prod_cat_df["prod_cat_code"].nunique()

6

In [8]:
trans_df.isnull().sum()

transaction_id      0
cust_id             0
tran_date           0
prod_subcat_code    0
prod_cat_code       0
Qty                 0
Rate                0
Tax                 0
total_amt           0
Store_type          0
dtype: int64

In [9]:
customer_df.head(3)

,customer_Id,DOB,Gender,city_code
0,268408,02-01-1970,M,4.0
1,269696,07-01-1970,F,8.0
2,268159,08-01-1970,F,8.0


In [16]:
prod_cat_df.head(3)

,prod_cat_code,prod_cat,prod_sub_cat_code,prod_subcat
0,1,Clothing,4,Mens
1,1,Clothing,1,Women
2,1,Clothing,3,Kids


In [13]:
trans_df.head(3)

,transaction_id,cust_id,tran_date,prod_subcat_code,prod_cat_code,Qty,Rate,Tax,total_amt,Store_type
0,80712190438,270351,28-02-2014,1,1,-5,-772,405.300,-4265.300,e-Shop
1,29258453508,270384,27-02-2014,5,3,-5,-1497,785.925,-8270.925,e-Shop
2,51750724947,273420,24-02-2014,6,5,-2,-791,166.110,-1748.110,TeleShop


In [14]:
trans_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23053 entries, 0 to 23052
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   transaction_id    23053 non-null  int64  
 1   cust_id           23053 non-null  int64  
 2   tran_date         23053 non-null  object 
 3   prod_subcat_code  23053 non-null  int64  
 4   prod_cat_code     23053 non-null  int64  
 5   Qty               23053 non-null  int64  
 6   Rate              23053 non-null  int64  
 7   Tax               23053 non-null  float64
 8   total_amt         23053 non-null  float64
 9   Store_type        23053 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.8+ MB


In [23]:
trans_df["transaction_id"].value_counts().head()

transaction_id
4170892941     4
32263938079    4
426787191      4
91377906980    3
44125492691    3
Name: count, dtype: int64

In [26]:
print("Minimum transaction date:", trans_df["tran_date"].min())
print("Maximum transaction date:", trans_df["tran_date"].max())

Minimum transaction date: 1/1/2012
Maximum transaction date: 9/9/2013


In [27]:
print("Number of negative quantities:", (trans_df["Qty"] < 0).sum())
print("Number of negative rates:", (trans_df["Rate"] < 0).sum())
print("Number of negative total amounts:", (trans_df["total_amt"] < 0).sum())

Number of negative quantities: 2177
Number of negative rates: 2177
Number of negative total amounts: 2177


In [29]:
trans_df[trans_df["Qty"] < 0].head(10)

,transaction_id,cust_id,tran_date,prod_subcat_code,prod_cat_code,Qty,Rate,Tax,total_amt,Store_type
0,80712190438,270351,28-02-2014,1,1,-5,-772,405.300,-4265.300,e-Shop
1,29258453508,270384,27-02-2014,5,3,-5,-1497,785.925,-8270.925,e-Shop
2,51750724947,273420,24-02-2014,6,5,-2,-791,166.110,-1748.110,TeleShop
3,93274880719,271509,24-02-2014,11,6,-3,-1363,429.345,-4518.345,e-Shop
4,51750724947,273420,23-02-2014,6,5,-2,-791,166.110,-1748.110,TeleShop
5,97439039119,272357,23-02-2014,8,3,-2,-824,173.040,-1821.040,TeleShop
6,45649838090,273667,22-02-2014,11,6,-1,-1450,152.250,-1602.250,e-Shop
7,22643667930,271489,22-02-2014,12,6,-1,-1225,128.625,-1353.625,TeleShop
8,79792372943,275108,22-02-2014,3,1,-3,-908,286.020,-3010.020,MBR
9,50076728598,269014,21-02-2014,8,3,-4,-581,244.020,-2568.020,e-Shop


In [39]:
print("Max number of unique customers per transaction ID:", trans_df.groupby("transaction_id")["cust_id"].nunique().max())

Max number of unique customers per transaction ID: 1


In [38]:
print("Max number of transaction dates per transaction ID:", trans_df.groupby("transaction_id")["tran_date"].nunique().max())

Max number of transaction dates per transaction ID: 4


In [45]:
trans_df.groupby("transaction_id")["tran_date"].nunique().value_counts()

tran_date
1    18821
2     1953
3      103
4        1
Name: count, dtype: int64

In [40]:
print("how many transaction IDs have multiple rows:", trans_df["transaction_id"].value_counts().gt(1).sum())

how many transaction IDs have multiple rows: 2057


In [41]:
trans_df["calculated_amt"] = (
    trans_df["Qty"] * trans_df["Rate"]
    + trans_df["Tax"]
)

trans_df["difference"] = (
    trans_df["calculated_amt"]
    - trans_df["total_amt"]
)

In [ ]:
trans_df.head(3)

,transaction_id,cust_id,tran_date,prod_subcat_code,prod_cat_code,Qty,Rate,Tax,total_amt,Store_type,calculated_amt,difference
0,80712190438,270351,28-02-2014,1,1,-5,-772,405.300,-4265.300,e-Shop,4265.300,8530.60
1,29258453508,270384,27-02-2014,5,3,-5,-1497,785.925,-8270.925,e-Shop,8270.925,16541.85
2,51750724947,273420,24-02-2014,6,5,-2,-791,166.110,-1748.110,TeleShop,1748.110,3496.22


### Current Discovery About the Datasets
```
CUSTOMER  
5,647 rows  
   │  
   │ customer_Id  
   ▼  
TRANSACTIONS  
23,053 rows  
20,878 transaction IDs  
   │  
   │ prod_cat_code +  
   │ prod_sub_cat_code  
   ▼  
PRODUCT CATEGORY  
23 rows  
```
#### Customer
1. 5,647 customers
2. 4 missing attribute values
3. 10 city codes
4. DOB approximately 1972–1992
#### Transactions
1. 23,053 rows
2. 20,878 unique transaction IDs
3. 2,057 transaction IDs have multiple rows
4. One customer per transaction ID
5. Some transaction IDs span multiple dates
6. 2,177 negative transaction lines
7. Negative Qty, Rate, and total_amt occur together
8. Financial relationship works for positive transactions
9. tran_date has inconsistent source formats

#### product_cat_info
```
prod_cat_code
      ↓
prod_cat
      ↓
prod_sub_cat_code + prod_cat_code
      ↓
prod_subcat
```